compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            suggestions.push((second_span, first_snippet));
+                        }
+                        Issue::Permutation(args) => {
+                            for (src, &arg) in args.iter().enumerate() {
+                                if let Some(dst) = arg {
+                                    let src_span = provided_args[src].span;
+                                    let dst_span = provided_args[dst].span;
+                                    let snippet = source_map.span_to_snippet(src_span).unwrap();
+                                    let expected_type = expected_input_tys[dst];
+                                    let found_type = final_arg_types.iter().filter(|(i, _, _)| *i == dst).next().unwrap().1;
+                                    labels.push((dst_span, format!("expected {}, found {}", expected_type, found_type)));
+                                    suggestions.push((dst_span, snippet));
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Now construct our error from the various things we've labeled
+                let highlight_sp = MultiSpan::from_spans(labels.iter().map(|s| s.0).collect());
+                let mut err = struct_span_err!(
+                    tcx.sess,
+                    highlight_sp,
+                    E0059, // FIXME: Choose a different code?
+                    "multiple arguments to this function are incorrect"
+                );
+
+                // Call out where the function is defined
+                if let Some(def_id) = fn_def_id {
+                    if let Some(node) = tcx.hir().get_if_local(def_id) {
+                        let mut spans: MultiSpan = node
+                            .ident()
+                            .map(|ident| ident.span)
+                            .unwrap_or_else(|| tcx.hir().span(node.hir_id().unwrap()))
+                            .into();
+
+                        if let Some(id) = node.body_id() {
+                            let body = tcx.hir().body(id);
+                            for param in body.params {
+                                spans.push_span_label(param.span, String::new());
+                            }
+                        }
+
+                        let def_kind = tcx.def_kind(def_id);
+                        err.span_note(spans, &format!("{} defined here", def_kind.descr(def_id)));
+                    }
+                }
+
+                // annotate each of the labels
+                for (span, label) in labels {
+                    err.span_label(span, label);
+                }
+
+                // And add a series of suggestions
+                // FIXME: for simpler cases, this might be overkill
+                if suggestions.len() > 0 {
+                    err.multipart_suggestion(
+                        "the following changes might help",
```

Comment:
It would be nice to say "swap the arguments" or "add the arguments" instead of just "changes".